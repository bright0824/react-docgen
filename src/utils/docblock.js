/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

/**
 * Helper functions to work with docblock comments.
 */

var DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gmi;

function parseDocblock(str) {
  var lines = str.split('\n');
  for (var i = 0, l = lines.length; i < l; i++) {
    lines[i] = lines[i].replace(/^\s*\*\s?/, '');
  }
  return lines.join('\n').trim();
}

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath): ?string {
  if (path.node.comments) {
    var comments = path.node.comments.filter(function(comment) {
      return comment.leading &&
        comment.type === 'CommentBlock' &&
        comment.value.indexOf('*\n') === 0;
    });
    if (comments.length > 0) {
      return parseDocblock(comments[comments.length - 1].value);
    }
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Object {
  var doclets = Object.create(null);
  var match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
