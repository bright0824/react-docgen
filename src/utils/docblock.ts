/**
 * Helper functions to work with docblock comments.
 */

import { CommentKind } from 'ast-types/gen/kinds';
import type { NodePath } from 'ast-types/lib/node-path';

const DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gim;

function parseDocblock(str: string): string {
  // Does not use \s in the regex as this would match also \n and conflicts
  // with windows line endings.
  return str.replace(/^[ \t]*\*[ \t]?/gm, '').trim();
}

const DOCBLOCK_HEADER = /^\*\s/;

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath, trailing = false): string | null {
  let comments: CommentKind[] = [];
  if (trailing && path.node.trailingComments) {
    comments = path.node.trailingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.leadingComments) {
    comments = path.node.leadingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.comments) {
    comments = path.node.comments.filter(
      comment =>
        comment.leading &&
        comment.type === 'CommentBlock' &&
        DOCBLOCK_HEADER.test(comment.value),
    );
  }

  if (comments.length > 0) {
    return parseDocblock(comments[comments.length - 1].value);
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Record<string, unknown> {
  const doclets = Object.create(null);
  let match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
