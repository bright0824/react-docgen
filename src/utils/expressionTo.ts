/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';
import type { NodePath } from 'ast-types/lib/node-path';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): string[] {
  const parts = [path];
  let result: string[] = [];

  while (parts.length > 0) {
    path = parts.shift() as NodePath;
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        // @ts-ignore
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      // @ts-ignore
      result.push(node.raw);
      continue;
    } else if (t.FunctionExpression.check(node)) {
      result.push('<function>');
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        if (t.SpreadElement.check(property.node)) {
          return `...${toString(property.get('argument'), importer)}`;
        } else {
          return (
            toString(property.get('key'), importer) +
            ': ' +
            toString(property.get('value'), importer)
          );
        }
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
