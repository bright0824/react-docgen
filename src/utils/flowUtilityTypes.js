/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (types.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return Boolean(idPath) && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  return path.get('typeParameters', 'params', 0);
}
