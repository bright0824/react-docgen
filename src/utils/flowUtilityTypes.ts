import { namedTypes as t } from 'ast-types';
import type { NodePath } from 'ast-types/lib/node-path';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  while (isSupportedUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}
