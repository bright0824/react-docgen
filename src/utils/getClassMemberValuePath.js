/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import getNameOrValue from './getNameOrValue';
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

export default function getClassMemberValuePath(
  classDefinition: NodePath,
  memberName: string,
): ?NodePath {
  // Fortunately it seems like that all members of a class body, be it
  // ClassProperty or MethodDefinition, have the same structure: They have a
  // "key" and a "value"
  return classDefinition
    .get('body', 'body')
    .filter(
      memberPath =>
        (!memberPath.node.computed ||
          types.Literal.check(memberPath.node.key)) &&
        !types.PrivateName.check(memberPath.node.key) &&
        getNameOrValue(memberPath.get('key')) === memberName &&
        memberPath.node.kind !== 'set',
    )
    .map(memberPath => memberPath.get('value'))[0];
}
