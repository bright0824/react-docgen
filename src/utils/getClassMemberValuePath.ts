import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  // Fortunately it seems like that all members of a class body, be it
  // ClassProperty or ClassMethod, have the same structure: They have a
  // "key" and a "value"
  return (
    classDefinition
      .get('body')
      .get('body')
      .filter(memberPath => {
        if (
          (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
          memberPath.isClassProperty()
        ) {
          const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
            'key',
          );
          return (
            (!memberPath.node.computed || key.isLiteral()) &&
            getNameOrValue(key) === memberName
          );
        }

        return false;
      }) //TODO ClassMethod does not have value
      .map(memberPath =>
        memberPath.isClassMethod()
          ? memberPath
          : (memberPath.get('value') as NodePath<Expression>),
      )[0] || null
  );
}
