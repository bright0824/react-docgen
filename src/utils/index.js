/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import * as docblock from './docblock';

export {
  docblock,
};

export {default as getClassMemberValuePath} from './getClassMemberValuePath';
export {default as getMemberExpressionRoot} from './getMemberExpressionRoot';
export {default as getMembers} from './getMembers';
export {default as getMemberValuePath} from './getMemberValuePath';
export {default as getNameOrValue} from './getNameOrValue';
export {default as getPropertName} from './getPropertyName';
export {default as getPropertyValuePath} from './getPropertyValuePath';
export {default as getPropType} from './getPropType';
export {default as isExportsOrModuleAssignment} from './isExportsOrModuleAssignment';
export {default as isReactComponentClass} from './isReactComponentClass.js';
export {default as isReactCreateClassCall} from './isReactCreateClassCall';
export {default as isReactModuleName} from './isReactModuleName';
export {default as match} from './match';
export {default as normalizeClassDefiniton} from './normalizeClassDefinition';
export {default as printValue} from './printValue';
export {default as resolveExportDeclaration} from './resolveExportDeclaration';
export {default as resolveToModule} from './resolveToModule';
export {default as resolveToValue} from './resolveToValue';
