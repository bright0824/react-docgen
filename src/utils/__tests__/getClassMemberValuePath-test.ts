import getClassMemberValuePath from '../getClassMemberValuePath';
import { parse } from '../../../tests/utils';
import type { ClassDeclaration, ClassExpression } from '@babel/types';

describe('getClassMemberValuePath', () => {
  describe('ClassDeclaration', () => {
    describe('MethodDefinitions', () => {
      it('finds "normal" method definitions', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          render() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
      });

      it('finds computed method definitions with literal keys', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          ['render']() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
      });

      it('ignores computed method definitions with expression', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          [render]() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'render')).toBeNull();
      });
    });

    describe('Getters and Setters', () => {
      it('finds getters', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          get foo() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toMatchSnapshot();
      });

      it('ignores setters', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          set foo(val) {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });
    });

    describe('ClassProperty', () => {
      it('finds "normal" class properties', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toEqual(
          def.get('body').get('body')[0].get('value'),
        );
      });
    });

    describe('private', () => {
      it('ignores private class properties', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });

      it('ignores private class methods', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #foo() {}
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBeNull();
      });

      it('finds "normal" class properties with private present', () => {
        const def = parse.statement<ClassDeclaration>(`
        class Foo {
          #private = 54;
          foo = 42;
        }
      `);

        expect(getClassMemberValuePath(def, 'foo')).toBe(
          def.get('body').get('body')[1].get('value'),
        );
      });
    });
  });

  describe('ClassExpression', () => {
    it('finds "normal" method definitions', () => {
      const def = parse.expression<ClassExpression>(`
      class Foo {
        render() {}
      }
    `);

      expect(getClassMemberValuePath(def, 'render')).toMatchSnapshot();
    });
  });
});
