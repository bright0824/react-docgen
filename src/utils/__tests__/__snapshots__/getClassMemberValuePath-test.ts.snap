// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClassMemberValuePath ClassDeclaration Getters and Setters finds getters 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": Array [],
    "directives": Array [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "foo",
    "type": "Identifier",
  },
  "kind": "get",
  "params": Array [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath ClassDeclaration MethodDefinitions finds "normal" method definitions 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": Array [],
    "directives": Array [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "render",
    "type": "Identifier",
  },
  "kind": "method",
  "params": Array [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath ClassDeclaration MethodDefinitions finds computed method definitions with literal keys 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": Array [],
    "directives": Array [],
    "type": "BlockStatement",
  },
  "computed": true,
  "generator": false,
  "id": null,
  "key": Node {
    "extra": Object {
      "raw": "'render'",
      "rawValue": "render",
    },
    "type": "StringLiteral",
    "value": "render",
  },
  "kind": "method",
  "params": Array [],
  "static": false,
  "type": "ClassMethod",
}
`;

exports[`getClassMemberValuePath ClassExpression finds "normal" method definitions 1`] = `
Node {
  "async": false,
  "body": Node {
    "body": Array [],
    "directives": Array [],
    "type": "BlockStatement",
  },
  "computed": false,
  "generator": false,
  "id": null,
  "key": Node {
    "name": "render",
    "type": "Identifier",
  },
  "kind": "method",
  "params": Array [],
  "static": false,
  "type": "ClassMethod",
}
`;
